File: TopKFrequentWords/top_k_frequent_words.dart
/*


 -* Top K Frequent Words *-

 Given an array of strings words and an integer k, return the k most frequent strings.

Return the answer sorted by the frequency from highest to lowest. Sort the words with the same frequency by their lexicographical order.



Example 1:

Input: words = ["i","love","leetcode","i","love","coding"], k = 2
Output: ["i","love"]
Explanation: "i" and "love" are the two most frequent words.
Note that "i" comes before "love" due to a lower alphabetical order.
Example 2:

Input: words = ["the","day","is","sunny","the","the","the","sunny","is","is"], k = 4
Output: ["the","is","sunny","day"]
Explanation: "the", "is", "sunny" and "day" are the four most frequent words, with the number of occurrence being 4, 3, 2 and 1 respectively.


Constraints:

1 <= words.length <= 500
1 <= words[i].length <= 10
words[i] consists of lowercase English letters.
k is in the range [1, The number of unique words[i]]


Follow-up: Could you solve it in O(n log(k)) time and O(n) extra space?


*/

class A {
  List<String> topKFrequent(List<String> words, int k) {
    final hashMap = <String, int>{};

    for (final word in words) {
      if (hashMap[word] == null) {
        hashMap[word] = 1;
      } else {
        hashMap[word] = hashMap[word]! + 1;
      }
    }

    final sortedHashMap = hashMap.entries.toList()
      ..sort((a, b) {
        if (a.value.compareTo(b.value) == 0) {
          return a.key.compareTo(b.key);
        }
        return b.value.compareTo(a.value);
      });

    return sortedHashMap.sublist(0, k).map((e) => e.key).toList();
  }
}

File: TopKFrequentWords/top_k_frequent_words.go
package main

import "container/heap"

type Word struct {
	word      string
	frequency int
}

func newWord(word string, frequency int) *Word {
	return &Word{word: word, frequency: frequency}
}

type H struct {
	items []*Word
}

func (h *H) Len() int { return len(h.items) }

func (h *H) Less(i, j int) bool {
	if h.items[i].frequency != h.items[j].frequency {
		return h.items[i].frequency < h.items[j].frequency
	}
	return h.items[i].word > h.items[j].word
}

func (h *H) Swap(i, j int) { h.items[i], h.items[j] = h.items[j], h.items[i] }

func (h *H) Push(i interface{}) { h.items = append(h.items, i.(*Word)) }

func (h *H) Pop() interface{} {
	n := len(h.items)
	item := h.items[n-1]
	h.items = h.items[:n-1]
	return item
}

func topKFrequent(words []string, k int) []string {
	counts := map[string]int{}
	h := &H{}
	for _, v := range words {
		counts[v]++

	}
	for key, v := range counts {
		heap.Push(h, newWord(key, v))
		if h.Len() > k {
			heap.Pop(h)
		}
	}
	result := []string{}
	for h.Len() > 0 {
		result = append([]string{heap.Pop(h).(*Word).word}, result...)
	}
	return result
}
